p2p_wallet/Scenes/Main/History/History.SourceStream.swift

```diff
@@ -0,0 +1,190 @@
+                Task {
+                    do {
+                        for source in sources {
+                            for try await transaction in source.next(configuration: configuration) {
+                                stream.yield(transaction)
+                            }
+                        }
+                        stream.finish(throwing: nil)
+                    } catch {
+                        stream.finish(throwing: error)
+                    }
+                }
+            }
+        }
+
+        func reset() {
+            for source in sources { source.reset() }
+        }
+    }
+
+    /// The class helps to retrieves all transactions as stream from defined account.
+    class AccountStreamSource: StreamSource {
+        let transactionRepository: TransactionRepository
+
+        /// The object that is responsible for parsing transactions
+        let transactionParser: TransactionParser
+
+        /// The account address
+        private let account: String
+
+        /// The account's token symbol
+        private let accountSymbol: String
+
+        /// The most latest signature of transactions, that has been loaded.
+        /// This value will be used as pagination indicator and all next transactions after this one will be loaded.
+        private var latestFetchedSignature: String?
+
+        /// Fixed number of transactions that will be requested each time.
+        private let batchSize: Int = 10
+
+        init(
+            account: String,
+            accountSymbol: String,
+            transactionRepository: TransactionRepository,
+            transactionParser: TransactionParser
+        ) {
+            self.account = account
+            self.accountSymbol = accountSymbol
+            self.transactionRepository = transactionRepository
+            self.transactionParser = transactionParser
+        }
+
+        func first() async throws -> SolanaSDK.ParsedTransaction? {
+            guard let signatureInfo = try await transactionRepository.getSignatures(
+                address: account,
+                limit: batchSize,
+                before: latestFetchedSignature
+            ).first else { return nil }
+
+            let transactionInfo = try await transactionRepository.getTransaction(signature: signatureInfo.signature)
+
+            return try await transactionParser.parse(
+                signatureInfo: signatureInfo,
+                transactionInfo: transactionInfo,
+                account: account,
+                symbol: accountSymbol
+            )
+        }
+
+        func next(configuration: FetchingConfiguration) -> AsyncThrowingStream<SolanaSDK.ParsedTransaction, Error> {
+            AsyncThrowingStream<SolanaSDK.ParsedTransaction, Error> { stream in
+                Task {
+                    do {
+                        while true {
+                            // Fetch transaction signatures
```

Comment:
so much nesting elements. After `while true` move next blocks into new func

Reply:
You are right about too much nesting elements. I have tried to break t he code but it will make code more complicated, since we use here `for try await` syntax and early breaking `if receivedItem > 10 { return }` to break stream. 

What we can do it combine AsyncStream with Task.
```swift
AsyncTaskStream { stream in 
  var receivedItem = 0
  while true {
     // More code here
  }
} 

```